### ISSUE TITLE
[Idea] Grandparents

### ISSUE THREAD

**Issue by PRabahy** [evolvio/evolv.io#51]
I'm thinking about making each creature inherit some traits from their grandparents. For example, its parents would contribute 90% of a new creatures traits and a grandparents would contribute 10%. I might end up making it some sort of exponential decay function (parents = 90%, grandparents = 9%, great-grandparents = .9%, etc).

**Comment by SirSKillz** [comment]
Can you explain your thinking some more? Each creature as it stands now would technically be part of their grandparent, so what ways are you thinking would make this different? Also, from an efficiency standpoint, creating code that has to keep track of multiple generations and then edit each new creature will probably slow things down significantly, depending on how you implemented them.

**Comment by PRabahy** [comment]
My thought was to try to emulate something sort of like recessive traits. A child creature might not be very fit due to random chance, but it might have had 2 very fit grandparents. I'm not sure if this will make much of a difference in the long run, but I curious if it will more quickly lead to more fit generations.

Another feature that I want to implement is the ability to visualize a family tree. Right now evolv.io shows the names of a creatures parents, but I think it would be really cool if you could click those names to actually show them on the map. Eventually there could be a different screen that would show the axons darkening, lightening, or moving in one creature to its offspring. I would love to be able to visualize something like a water avoidance trait being developed.

**Comment by SirSKillz** [comment]
In terms of evolution, it wouldn't make sense to give traits from the grandparent to the offspring, because they should already be part of the child. The code doesn't randomly assign traits to the offspring, it's based on the parents (and through this, the grandparents) and then a small chance that mutation will occur and change the child to be more or less fit for the current environment. So technically, the child should already have the traits for the grandparents, unless it mutated. And if it mutated and you changed the mutation back to the original, then that would slow evolution severely because it would have to both mutate and then get passed inheriting characteristics from its grandparents. The wonderful thing about evolution is that there are many ways to "prepare the egg" such as avoiding water or just going very fast over water, so it doesn't really make sense to decide which is better for the creature; the mutations are there to cause evolution, so reverting to the past wouldn't make sense, at least not to me. 

In terms of the visual family tree, I think that would be a very interesting feature as well, and make it much easier to see evolution occurring. It would be fantastic to be able to see how evolution has taken place, especially after many, many, many generations. Are you thinking of a separate program that would read a save file on the generations, or something that could run and update at the same time as the simulator? Or we could have both, maybe the after action report that allows us to very deeply examine each creature and show us exactly how everything started, and the simultaneous part give a broad overview of the generations?

What are your thoughts, both on the grandparents and the family tree?

**Comment by PRabahy** [comment]
The grandparents thing would be more of an experiment just to see how it effects the evolution.

I like both ideas for implementing the family tree. I was thinking of just having it as part of the main program, but you could do some deeper analysis if it was in a post processing step.

I have been thinking of ways to classify creatures into species which would be pretty neat. Each creature could be give a certain number of points. Each point could be used for one action (using sight, turning, accelerating, eating, etc.). When an action is preformed, the neural net would decide how much (-1 to 1) effort to put into that action. A creature could evolve to do above 1 of an action by assigning 2 points to that action. With this, the number of points to each action become a species (1Sight2Turn4Accel2Eat -> 1S2T4A2E). (Hopefully that all made sense)

If we end up saving an entire family tree, we could do something like find a creature that died and bring it back to life. Natural evolution is cool, but some times it is more fun to play god.

**Comment by SirSKillz** [comment]
Alright. Well seeing as the grandparent thing is more of a creative idea, I say let's keep it out of the master for now (Maybe plugins in the future?)

The family tree thing would be very cool, but I think the easy way of creating species would be using the family tree to determine the family. People who stem from so-and-so group could be named using the first several letters of it's founders name, which would make coding for it significantly easier. A different approach, if you wanted to use your point system, could be a naming system similar to what you brought up, so we would find the max and min of the entire population, and the have intervals between each which would separate up the creatures into groups, and then the species would be ones in the same category for each topic (sight, turn, acceleration, eat, in your case).

Also, how familiar are you with the process of evolving? In a few of your messages, I get confused because you are saying things that don't happen in evolution, so I have to make a call as to what you mean. Not to offend you, I just want to make sure I'm understanding what you are trying to say.

**Comment by PRabahy** [comment]
To answer your last question, I'm just glanced at the actual evolution code
a little bit. I understand the concept quite well and get the general idea
that this implementation is hoing for. I'm going to be away from my
computer for a few days, but I'm working on a pretty major overhaul to the
creature/brain classes. I made a CreatureAction interface and am moving
each action into its own class that implements the interface. Each creature
has a list of CreatureActions that it preforms each turn. This ends up
adding another entire dimension that the creature can evolve while at the
same time cleaning up the code and making it more extensible. When I'm
done, creatures will be able to do something like grow a new eye as an
evolution.

**Comment by SirSKillz** [comment]
No, I'm not talking about the code that we are using. I'm talking about literal evolution, as in through science. Several times I have read your comments and taken it as the creature is deciding to do this, rather than the creature is randomly evolving over time. An example could be your last message; you said "When I'm done, creatures will be able to do something like grow a new eye as an evolution." Which is a possibility as it can happen, but your wording made me think that the creature was selecting traits it thinks will help it the most, rather than creatures having to deal with what is randomly assigned to them. If I were wording it, I probably would've said, "When I'm done, it would be possible for mutations such as new eyes to occur." Sorry to have to bring it up, but I'm confused and don't want you to work for something that is not evolution, so I would rather make sure now, than regret it later.

**Comment by PRabahy** [comment]
Ah, that was your concern. Yes I understand that each child creature is a
combination of its parents with some random mutations.

When I talk about code, I talk in an object oriented fashion. My point of
view for my statement was from inside the creature object, so I meant the
creature object will have a reproduce method that had the potential of
making a new creature object with an additional eye. Your interpretation of
that is exactly correct.

In concrete terms, if there is a creature with 2 eyes, it has something
like a 90% chance of having a child creature with 2 eyes, and a 10% chance
of having a child creature with 1 or 3 eyes. The exact percentages are
tuned by constants in the source code.

**Comment by SirSKillz** [comment]
Alright. My worries have been put to rest. Good luck!

**Comment by izik1** [comment]
Is this still active @PRabahy?

**Comment by PRabahy** [comment]
Yes-ish. I still think its a good idea, but I haven't done anything about it recently. For now I have been working on code cleanup and infrastructure. Before I dive too deeply into this, I want to make the brain have a dynamic number of input and outputs.
